featurebyte/service/feature_materialize.py

```diff
@@ -0,0 +1,332 @@
+from featurebyte.query_graph.node.schema import TableDetails
+from featurebyte.query_graph.sql.adapter import get_sql_adapter
+from featurebyte.query_graph.sql.common import quoted_identifier, sql_to_string
+from featurebyte.query_graph.sql.online_serving import (
+    TemporaryBatchRequestTable,
+    get_online_features,
+)
+from featurebyte.service.entity import EntityService
+from featurebyte.service.feature import FeatureService
+from featurebyte.service.online_store_compute_query_service import OnlineStoreComputeQueryService
+from featurebyte.service.online_store_table_version import OnlineStoreTableVersionService
+from featurebyte.session.base import BaseSession
+
+
+@dataclass
+class OfflineIngestGraphMetadata:
+    """
+    Information about offline ingest graph combined for all features
+    """
+
+    graph: QueryGraph
+    node_names: List[str]
+    output_column_names: List[str]
+
+
+@dataclass
+class OnlineEnabledFeaturesMetadata:
+    """
+    Information about online enabled features having the same primary entity that should be
+    published together
+    """
+
+    features: List[FeatureModel]
+    online_store_table_names: List[str]
+    primary_entities: List[EntityModel]
+    ingest_graph_and_node_names: Tuple[QueryGraph, List[str]] = field(init=False)
+    output_column_names: List[str] = field(init=False)
+
+    def __post_init__(self) -> None:
+        ingest_graph_metadata = self._get_combined_ingest_graph()
+        self.ingest_graph_and_node_names = (
+            ingest_graph_metadata.graph,
+            ingest_graph_metadata.node_names,
+        )
+        self.output_column_names = ingest_graph_metadata.output_column_names
+
+    @property
+    def serving_names(self) -> List[str]:
+        """
+        Returns serving names for the primary entities
+
+        Returns
+        -------
+        List[str]
+        """
+        return sorted([entity.serving_names[0] for entity in self.primary_entities])
+
+    @property
+    def primary_entity_ids(self) -> List[ObjectId]:
+        """
+        Returns primary entity ids
+
+        Returns
+        -------
+        List[ObjectId]
+        """
+        return sorted([entity.id for entity in self.primary_entities])
+
+    @property
+    def entity_universe_expr(self) -> Optional[Select]:
+        """
+        Returns a SQL expression that selects the distinct entity serving names across all online
+        store tables
+
+        Returns
+        -------
+        Optional[Select]
+        """
+        if not self.online_store_table_names:
+            return None
+
+        # select distinct serving names across all online store tables
+        distinct_serving_names_from_tables = [
+            select(*[quoted_identifier(serving_name) for serving_name in self.serving_names])
+            .distinct()
+            .from_(table_name)
+            for table_name in self.online_store_table_names
+        ]
```

Comment:
Just to double check my understanding, the online store table here refers to offline store table in feast. What we are going to do here is to get the entity universe of all the offline store tables

Reply:
The online store tables here refer to our current online store tables, not the ones in feast. Currently the materialisation is based on `get_online_features`, which is batch materialisation using precomputed aggregation results.